apps/files/helpers.py

```diff
@@ -335,30 +342,143 @@ def get_url(self, short):
+                        left_files[key]['diff'] = True
+
+            self._files = left_files
+        return self._files
+
+    @memoize(prefix='file-viewer-remap-files', time=60 * 60)
+    def _remap_filenames(self):
+        """
+        Remap the filenames of add-ons whose directory contents have
+        been re-arranged. Currently handles SDK add-ons moving from
+        the `cfx` packaging format to the `jpm` format.
+        """
+        left = self.left.files
+        right = self.right.files
+
+        if u'harness-options.json' not in right or u'package.json' not in left:
+            return right
+
+        try:
+            with storage.open(right[u'harness-options.json']['full']) as f:
+                harness_options = json.load(f)
+
+            with storage.open(left['package.json']['full']) as f:
+                package = json.load(f)
+
+            assert isinstance(harness_options['mainPath'], basestring)
+            assert '/' in harness_options['mainPath']
+        except:
+            # Any errors here, just bail. Something is malformed in
+            # the add-on, and there's nothing we can do about it.
+            return right
+
+        main = '%s.js' % harness_options['mainPath'].replace('/', '/lib/')
+        new_main = package.get('main', 'index.js')
+        main_package = main[:main.index('/')]
+
+        have_modules = left.get('node_modules', {}).get('directory')
+        if have_modules:
+            # Need to create a 'node_modules' directory in the right
+            # file tree, just to try to ward off any possible
+            # issues. Easiest thing is to steal it from the left
+            # side.
+            right['node_modules'] = dict(left['node_modules'])
+
+        def move(src, dst):
+            if dst not in right:
+                right[dst] = right[src]
+                del right[src]
+
+        package = None
+        prefix = '/'
+        new_prefix = None
+
+        # No iterators here. We're going to be modifying the dict,
+        # so they'll break.
+        for path in right.keys():
+            if path.startswith('resources/'):
+                p = path[len('resources/'):]
+
+                if p == main:
+                    # Always try to move the old main module to the
+                    # location of the new main, regardless of
+                    # package.
+                    move(path, new_main)
+                elif p.startswith('addon-sdk/') or p == 'addon-sdk':
+                    # No longer used. Just drop.
+                    del right[path]
+                elif '/' not in p:
+                    # Package directory
+                    package = p
+                    prefix = '%s/' % p
+                    if package == main_package:
+                        new_prefix = ''
+                    else:
+                        if have_modules:
+                            new_prefix = 'node_modules/%s/' % p
+                        else:
```

Comment:
You can get rid of this `else` by changing line 442 to `prefix = new_prefix = '%s/' % p`